py-polars/polars/expr/expr.py

```diff
@@ -8738,274 +8730,9 @@ def map_dict(
-                        right_on=remap_key_column,
-                    )
-                    .select(
-                        F.when(F.col(is_remapped_column).is_not_null())
-                        .then(F.col(remap_value_column))
-                        .otherwise(default_parsed)
-                        .alias(column)
-                    )
-                )
-                .collect(no_optimization=True)
-                .to_series()
-            )
-
-        def inner(s: Series) -> Series:
-            column = s.name
-            input_dtype = s.dtype
-            remap_key_column = f"__POLARS_REMAP_KEY_{column}"
-            remap_value_column = f"__POLARS_REMAP_VALUE_{column}"
-            is_remapped_column = f"__POLARS_REMAP_IS_REMAPPED_{column}"
-
-            remap_key_s = _remap_key_or_value_series(
-                name=remap_key_column,
-                values=list(remapping.keys()),
-                dtype=input_dtype,
-                dtype_if_empty=input_dtype,
-                dtype_keys=input_dtype,
-                is_keys=True,
-            )
-
-            if return_dtype:
-                # Create remap value Series with specified output dtype.
-                remap_value_s = pl.Series(
-                    remap_value_column,
-                    remapping.values(),
-                    dtype=return_dtype,
-                    dtype_if_empty=input_dtype,
-                )
-            else:
-                # Create remap value Series with same output dtype as remap key Series,
-                # if possible (if both are integers, both are floats or remap value
-                # Series is pl.Utf8 and remap key Series is pl.Categorical).
-                remap_value_s = _remap_key_or_value_series(
-                    name=remap_value_column,
-                    values=remapping.values(),
-                    dtype=None,
-                    dtype_if_empty=input_dtype,
-                    dtype_keys=input_dtype,
-                    is_keys=False,
-                )
-
-            return (
-                (
-                    s.to_frame()
-                    .lazy()
-                    .join(
-                        pl.DataFrame(
-                            [
-                                remap_key_s,
-                                remap_value_s,
-                            ]
-                        )
-                        .lazy()
-                        .with_columns(F.lit(True).alias(is_remapped_column)),
-                        how="left",
-                        left_on=column,
-                        right_on=remap_key_column,
-                    )
-                )
-                .collect(no_optimization=True)
-                .to_series(1)
-            )
-
-        func = inner_with_default if default is not None else inner
-        return self.map(func)
+        return operations.map_dict(
+            self, remapping, default=default, return_dtype=return_dtype
+        )
```

Comment:
Instead of dispatching to another function, you can simply import the function directly here, preferrably directly inside the class for clarity:

```python
from polars.expr.operations import map_dict
```

This will save one level of indirection.

Reply:
> It should just work.

It doesn't (just tried it), or I am misunderstanding what you're suggesting...

Reply:
```python
# my_class.py
class MyClass:

    def __init__(self):
        self._a = 1
        
    from _submodule import print_a
```

```python
# _submodule.py
def print_a(self):
    print(self._a)
```

```
>>> from my_class import MyClass
>>> x = MyClass()
>>> x.print_a()
1
```

Reply:
Oh wow, a whole new world just opened up for me :star_struck:

This does break the `@expr_dispatch` though, so we'd have to solve that somehow.

Reply:
Alternative, the following approach may be a bit cleaner, and I *think* possibly more performant in how Python resolves class/instance methods (@alexander-beedie can probably chime in here with more knowledge):

```python
# dataframe.py
from _submodules import ops

class DataFrame():
    def __init__(self):
        ...

    # assign methods directly to imports
    map_dict = ops.map_dict
    transpose = ops.transpose
    ...

Reply:
I strongly prefer this latter variant ;)  

Not because there is any performance difference in use (there isn't - the classes will perform _identically_), and not because of the (incredibly minor) overhead on initial class (but not instance) creation, but mainly because it's just vastly more obvious what's happening, heh.

Reply:
Ok, I agree now.

FYI, speaking of minor performance gains, should we look into using `__slots__` in our objects? I implemented this once and had to come up with a hacky workaround to namespace registrations (namely, had to hack the `__getattribute__` method so that when you call `df.<x>`, it dispatches `x` correctly, since you can't dynamically add attributes when `__slots__ is defined), which probably killed any performance gains.

Reply:
Probably not worth it, polars isn't used in a way that `df.<x>` is called millions of times.